### PR TITLE
 Test whether shelf list shows new shelf

### LINKED ISSUE #253: Test whether shelf list shows new shelf
After adding a new shelf, the shelf list in the 'my books' page (the dropdown above the book grid) should show the newly created shelf (a manual refresh should not be needed).

There was a bug (#244) that was recently fixed, so we would like for there to be a test to ensure that we are notified (by way of a failing test) if this happens in the future.

### PR DESCRIPTION
**Pull request by MikeLowrie**
## Summary of change

Test file created to verify that adding a new shelf updates the UI.

## Related issue

Closes #253 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-h6vx3n8l-KN8QnO50r7QWZHgHFnFdPw)

### CODE REVIEW

**Comment by MikeLowrie** [comment]
The comment on removing the getAllShelves() method in the BookShelfComboBox has been addressed, but is not present in the Files Changed tab.

File: src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.test.context.web.WebAppConfiguration;
+
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (MikeLowrie) [inline src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java:23]: assertThat static import created to allow static call for assert.

File: src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+@IntegrationTest
+@WebAppConfiguration
+class NewShelfTest {
```
Comment (MikeLowrie) [inline src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java:27]: `public` modifier removed.

File: src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * setup here has been copied from BooksInShelfViewTest, as this test requires a shelf view as well.
+     */
+    @BeforeEach
+    public void setup(@Autowired BookService bookService,
+                      @Autowired PredefinedShelfService predefinedShelfService,
+                      @Autowired CustomShelfService customShelfService) {
+        final SpringServlet servlet = new MockSpringServlet(routes, ctx);
+        MockVaadin.setup(UI::new, servlet);
+
```
Comment (MikeLowrie) [inline src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java:49]: `assumeTrue` call removed.

File: src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        this.customShelfService = customShelfService;
+        shelfView = new BooksInShelfView(bookService, predefinedShelfService, customShelfService);
+    }
+
+    /**
+     * This test verifies that when creating a new shelf and adding it to the shelf list,
+     * the dropdown that shows all shelves contains the new shelf.
+     */
```
Comment (MikeLowrie) [inline src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java:57]: Comments cleaned up to not include mention of `IllegalStateException`.

File: src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        USE A LISTDATAPROVIDER TO GET ITEMS FROM A ComboBox
+        REACH OUT TO KARAN FOR HELP
+         * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+         */
+
+        customShelfService.save(test);
+        shelfView.whichShelf.updateShelfList();
+
+        //////////////////////////////////////////////////////                                   ADD STATIC IMPORT FOR assertTrue      /////////////////
+        assertThat(customShelfService.findAll("UnitTest").size() > 0).isTrue();
```
Comment (MikeLowrie) [inline src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java:72]: Test needs to be adjusted to use a `ListDataProvider `and read from the ComboBox directly.
One thing I've noticed is the ComboBox does not have a `DataProvider `of any kind connected to it beforehand. Should that object be reconfigured to leverage a `DataProvider`?

Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/ui/shelf/NewShelfTest.java]: `comboBox.getDataProvider()` can be used

**Comment by knjk04** [comment]
@MikeLowrie Apologies, I didn't realise it was a bit tricky to retrieve the list of values from a `ComboBox`. To help you out, I've retrieved the list of all of the items from the list (I hope you don't mind).

I've moved things around a little. The test is now in `BookShelfComboBoxTest`, but it should still look familiar.

We still need to add the assertion. I've added a comment there as a tip, so please remove that when you're finished with the assertion. If you're not familiar with AssertJ, there's an elegant way to check whether a particular element is in a list ([AssertJ's documentation](https://assertj.github.io/doc/) is good).

As ever, feel free to reach out if you need a hand with this!

**Comment by MikeLowrie** [comment]
Thanks for the assistance on pulling data from a ComboBox. I was reviewing Vaadin's documentation on ComboBoxes and trying some different things but couldn't get anything to work correctly, so I appreciate the direction on your part.

**Review by knjk04 — APPROVED** [review]
Thanks for working on this! Good job on the assertion

**Comment by knjk04** [comment]
> Thanks for the assistance on pulling data from a ComboBox. I was reviewing Vaadin's documentation on ComboBoxes and trying some different things but couldn't get anything to work correctly, so I appreciate the direction on your part.

My pleasure!

**Pull request merged**